File: src/python/main/JSON/UpdateInfo.py
from collections import namedtuple
from src.python.main.JSON.JSONManager import JSONManager as JManager


class UpdateInfo(object):

    def __init__(self):
        # Data that's gonna be sent to the server
        self.playersName = []
        self.playersGameOver = []
        self.treeB = ""
        self.treeBST = ""
        self.treAVL = ""
        self.treeSplay = ""

        # Data that's gonna be received from the server
        self.time = 0
        self.tokenSend = ""
        self.challenge = []

    def updateWrapper(self, playersName, playersGameOver, treeB, treeBST, treeAVL, treeSplay):
        self.setPlayersName(playersName)
        self.setPlayersGameOver(playersGameOver)
        self.setTreeB(treeB)
        self.setTreeBST(treeBST)
        self.setTreeAVL(treeAVL)
        self.setTreeSplay(treeSplay)
        dictionary = self.asdict()
        return dictionary

    def asdict(self):
        return {"playersName": self.playersName, "playersGameOver": self.playersGameOver, "treeB": self.treeB,
                "treeBST": self.treeBST, "treeAVL": self.treeAVL, "treeSplay": self.treeSplay}

    def unwrapper(self, dictio):
        self.setTime(dictio['time'])
        self.setTokenSend(dictio['tokenSend'])
        self.setChallenge(dictio['challenge'])
        print(self.time)
        print(self.tokenSend)
        print(self.challenge)

    def setPlayersName(self, playersName):
        self.playersName = playersName

    def getPlayersName(self):
        return self.playersName

    def setPlayersGameOver(self, playersGameOver):
        self.playersGameOver = playersGameOver

    def getPlayersGameOver(self):
        return self.playersGameOver

    def setTreeB(self, treeB):
        self.treeB = treeB

    def getTreeB(self):
        return self.treeB

    def setTreeBST(self, treeBST):
        self.treeBST = treeBST

    def getTreeBST(self):
        return self.treeBST

    def setTreeAVL(self, treeAVL):
        self.treeAVL = treeAVL

    def getTreeAVL(self):
        return self.treeAVL

    def setTreeSplay(self, treeSplay):
        self.treeSplay = treeSplay

    def getTreeSplay(self):
        return self.treeSplay

    def setTime(self, time):
        self.time = time

    def getTime(self):
        return self.time

    def setTokenSend(self, tokenSend):
        self.tokenSend = tokenSend

    def getTokenSend(self):
        return self.tokenSend

    def setChallenge(self, challenge):
        self.challenge = challenge

    def getChallenge(self):
        return self.challenge


# chamuco = UpdateInfo()
# print(chamuco.updateWrapper(["LOLA", "LOLO"], [False, False], "", "treeBST@24", "", ""))

# ["LOLA", "LOLO"], [false, false], "", "treeBST@24", "", ""

#cogollo = UpdateInfo()
#cogollo.unwrapper(
#    {"playerName": [], "playerGameOver": [], "treeB": "", "treeBST": "", "treeAVL": "", "treeSplay": "", "time": 44,
#     "tokenSend": "treeAVL@88", "challenge": ["treeB", "treeAVL"]})

#Ramon = JManager().selectAction('{"playerName": [], "playerGameOver": [], "treeB": "", "treeBST": "", "treeAVL": "", "treeSplay": "", "time": "44","tokenSend": "treeAVL@88", "challenge": ["treeB", "treeAVL"]}')
#print(Ramon)
#desensablar = UpdateInfo()

#desensablar.unwrapper(Ramon)

la = UpdateInfo()
print (la.updateWrapper(["LOLA", "LOLO"], [False, False], "", "treeBST@24", "", ""))
lo = JManager().selectAction(la.updateWrapper(["LOLA", "LOLO"], [False, False], "", "treeBST@24", "", ""))

File: src/javas/conection/UpdateInfo.java
package javas.conection;

/**
 * Class that holds the info that must be updated and treated by de server
 */
public class UpdateInfo {
    /**
     * Instance of the Update info object
     */
    private static UpdateInfo instance;
    /**
     *List of Strings with the names of the players
     */
    private String[] playersName;
    /**
     *List of booleans with the current state of the players, true alive, false death
     */
    private boolean[] playersGameOver = {true, true, true, true};
    /**
     *Token grabbed by the player with the treeB assigned
     */
    private String treeB;
    /**
     *Token grabbed by the player with the treeBST assigned
     */
    private String treeBST;
    /**
     *Token grabbed by the player with the treeAVL assigned
     */
    private String treeAVL;
    /**
     *Token grabbed by the player with the treeSplay assigned
     */
    private String treeSplay;
    /**
     *String with the time value that the server is going to send to the client
     */
    private String time = "0:0";
    /**
     *Token that will be showed next on the client's interface
     */
    private String tokenSend = "";
    /**
     *List with the challenges settled for each player
     */
    private String[] challenge = {"", "", "", ""};

    private UpdateInfo(){

    }

    /**
     * This constructor works as a Singleton, it creates an instance and if the instance is created, it
     * returns it
     * @return instance
     */
    public static UpdateInfo getUpdateInfo(){
        if (instance == null){
            instance = new UpdateInfo();
        }
        return instance;
    }

    /**
     * This constructor updates the time value when game starts
     * @param time
     */
    public UpdateInfo(String time){
        this.time = time;
    }

    /**
     * This method updates the whole files info.
     * @param newInfo
     */
    public void UpdateFile(UpdateInfo newInfo){
        this.playersName = newInfo.getPlayersName();
        this.playersGameOver = newInfo.getPlayersGameOver();
        this.treeB = newInfo.getTreeB();
        this.treeBST = newInfo.getTreeBST();
        this.treeAVL = newInfo.getTreeAVL();
        this.treeSplay = newInfo.getTreeSplay();
    }

    /**
     * Sets new challenge.
     *
     * @param challenge New value of challenge.
     */
    public void setChallenge(String[] challenge) {
        this.challenge = challenge;
    }

    /**
     * Gets treeAVL.
     *
     * @return Value of treeAVL.
     */
    public String getTreeAVL() {
        return treeAVL;
    }

    /**
     * Gets time.
     *
     * @return Value of time.
     */
    public String getTime() {
        return time;
    }

    /**
     * Sets new treeAVL.
     *
     * @param treeAVL New value of treeAVL.
     */
    public void setTreeAVL(String treeAVL) {
        this.treeAVL = treeAVL;
    }

    /**
     * Sets new playersName.
     *
     * @param playersName New value of playersName.
     */
    public void setPlayersName(String[] playersName) {
        this.playersName = playersName;
    }

    /**
     * Sets new time.
     *
     * @param time New value of time.
     */
    public void setTime(String time) {
        this.time = time;
    }

    /**
     * Gets treeBST.
     *
     * @return Value of treeBST.
     */
    public String getTreeBST() {
        return treeBST;
    }

    /**
     * Gets playersName.
     *
     * @return Value of playersName.
     */
    public String[] getPlayersName() {
        return playersName;
    }

    /**
     * Sets new playersGameOver.
     *
     * @param playersGameOver New value of playersGameOver.
     */
    public void setPlayersGameOver(boolean[] playersGameOver) {
        this.playersGameOver = playersGameOver;
    }

    /**
     * Sets new treeSplay.
     *
     * @param treeSplay New value of treeSplay.
     */
    public void setTreeSplay(String treeSplay) {
        this.treeSplay = treeSplay;
    }

    /**
     * Gets treeB.
     *
     * @return Value of treeB.
     */
    public String getTreeB() {
        return treeB;
    }

    /**
     * Sets new treeB.
     *
     * @param treeB New value of treeB.
     */
    public void setTreeB(String treeB) {
        this.treeB = treeB;
    }

    /**
     * Gets playersGameOver.
     *
     * @return Value of playersGameOver.
     */
    public boolean[] getPlayersGameOver() {
        return playersGameOver;
    }

    /**
     * Sets new tokenSend.
     *
     * @param tokenSend New value of tokenSend.
     */
    public void setTokenSend(String tokenSend) {
        this.tokenSend = tokenSend;
    }

    /**
     * Gets treeSplay.
     *
     * @return Value of treeSplay.
     */
    public String getTreeSplay() {
        return treeSplay;
    }

    /**
     * Gets challenge.
     *
     * @return Value of challenge.
     */
    public String[] getChallenge() {
        return challenge;
    }

    /**
     * Sets new treeBST.
     *
     * @param treeBST New value of treeBST.
     */
    public void setTreeBST(String treeBST) {
        this.treeBST = treeBST;
    }

    /**
     * Gets tokenSend.
     *
     * @return Value of tokenSend.
     */
    public String getTokenSend() {
        return tokenSend;
    }
}
